File: src/mpg_leads/fetcher.py
import json
import time
from pathlib import Path

import requests

from mpg_leads.models import Business

PLACES_URL = "https://places.googleapis.com/v1/places:searchNearby"
PRICE_LEVELS = {
    "PRICE_LEVEL_FREE": 0, "PRICE_LEVEL_INEXPENSIVE": 1,
    "PRICE_LEVEL_MODERATE": 2, "PRICE_LEVEL_EXPENSIVE": 3,
    "PRICE_LEVEL_VERY_EXPENSIVE": 4,
}
FIELD_MASK = ",".join([
    "places.id", "places.displayName", "places.primaryType",
    "places.formattedAddress", "places.nationalPhoneNumber",
    "places.websiteUri", "places.rating", "places.userRatingCount",
    "places.priceLevel", "places.businessStatus", "places.reviews",
])

# Google returns granular primary types; map the common ones into our ICP vocab.
TYPE_NORMALIZATION = {
    "hair_salon": "salon", "beauty_salon": "salon", "nail_salon": "salon",
    "barber_shop": "salon", "spa": "spa", "day_spa": "spa",
    "restaurant": "restaurant", "meal_takeaway": "restaurant",
    "meal_delivery": "restaurant", "pizza_restaurant": "restaurant",
    "mexican_restaurant": "restaurant", "cafe": "cafe", "coffee_shop": "cafe",
    "bar": "bar", "pub": "bar", "night_club": "bar",
    "clothing_store": "retail", "store": "retail", "shoe_store": "retail",
    "gift_shop": "retail", "furniture_store": "retail", "boutique": "retail",
    "car_repair": "auto", "car_wash": "auto", "auto_parts_store": "auto",
    "dentist": "professional", "doctor": "professional", "lawyer": "professional",
    "accounting": "professional", "veterinary_care": "professional",
}


def normalize_category(primary_type: str) -> str:
    return TYPE_NORMALIZATION.get(primary_type, primary_type)


def parse_places_response(raw: dict) -> list[Business]:
    out = []
    for p in raw.get("places", []):
        reviews = [r.get("text", {}).get("text", "")
                   for r in p.get("reviews", []) if r.get("text")]
        out.append(Business(
            place_id=p["id"],
            name=p.get("displayName", {}).get("text", ""),
            category=normalize_category(p.get("primaryType", "")),
            address=p.get("formattedAddress", ""),
            phone=p.get("nationalPhoneNumber"),
            website=p.get("websiteUri"),
            rating=p.get("rating"),
            review_count=p.get("userRatingCount", 0),
            price_level=PRICE_LEVELS.get(p.get("priceLevel")),
            business_status=p.get("businessStatus", ""),
            review_texts=[t for t in reviews if t],
        ))
    return out


def dedupe(businesses: list[Business], seen_ids: set[str]) -> list[Business]:
    return [b for b in businesses if b.place_id not in seen_ids]


def load_seen(path: str) -> set[str]:
    p = Path(path)
    if not p.exists():
        return set()
    return set(json.loads(p.read_text()))


def save_seen(path: str, seen_ids: set[str]) -> None:
    Path(path).write_text(json.dumps(sorted(seen_ids)))


def load_demo_businesses() -> list[Business]:
    """Load bundled demo data so the dashboard renders without an API key."""
    fixture = Path(__file__).parent / "static" / "demo_places.json"
    return parse_places_response(json.loads(fixture.read_text()))


def fetch_nearby(api_key, location, radius_meters, included_types,
                 max_results=20, retries=3) -> list[Business]:
    """Call Places Nearby Search. Returns list[Business]. Retries on 429."""
    lat, lng = (float(x) for x in location.split(","))
    body = {
        "includedTypes": included_types,
        "maxResultCount": max_results,
        "locationRestriction": {"circle": {
            "center": {"latitude": lat, "longitude": lng},
            "radius": float(radius_meters)}},
    }
    headers = {
        "Content-Type": "application/json",
        "X-Goog-Api-Key": api_key,
        "X-Goog-FieldMask": FIELD_MASK,
    }
    for attempt in range(retries):
        resp = requests.post(PLACES_URL, headers=headers, json=body, timeout=30)
        if resp.status_code == 429:
            time.sleep(2 ** attempt)
            continue
        resp.raise_for_status()
        return parse_places_response(resp.json())
    raise RuntimeError("Places API rate limit: exhausted retries")

File: lib/fetcher.js
import { business } from "./models.js";
// Static JSON import so esbuild inlines the demo data into the function bundle
// (createRequire + a relative path is NOT bundle-safe on Netlify).
import demoRaw from "../public/demo_places.json" with { type: "json" };

export const PLACES_URL = "https://places.googleapis.com/v1/places:searchNearby";

export const PRICE_LEVELS = {
  PRICE_LEVEL_FREE: 0, PRICE_LEVEL_INEXPENSIVE: 1,
  PRICE_LEVEL_MODERATE: 2, PRICE_LEVEL_EXPENSIVE: 3,
  PRICE_LEVEL_VERY_EXPENSIVE: 4,
};

export const FIELD_MASK = [
  "places.id", "places.displayName", "places.primaryType",
  "places.formattedAddress", "places.nationalPhoneNumber",
  "places.websiteUri", "places.rating", "places.userRatingCount",
  "places.priceLevel", "places.businessStatus", "places.reviews",
].join(",");

const TYPE_NORMALIZATION = {
  hair_salon: "salon", beauty_salon: "salon", nail_salon: "salon",
  barber_shop: "salon", spa: "spa", day_spa: "spa",
  restaurant: "restaurant", meal_takeaway: "restaurant",
  meal_delivery: "restaurant", pizza_restaurant: "restaurant",
  mexican_restaurant: "restaurant", cafe: "cafe", coffee_shop: "cafe",
  bar: "bar", pub: "bar", night_club: "bar",
  clothing_store: "retail", store: "retail", shoe_store: "retail",
  gift_shop: "retail", furniture_store: "retail", boutique: "retail",
  car_repair: "auto", car_wash: "auto", auto_parts_store: "auto",
  dentist: "professional", doctor: "professional", lawyer: "professional",
  accounting: "professional", veterinary_care: "professional",
};

export function normalizeCategory(primaryType) {
  return TYPE_NORMALIZATION[primaryType] ?? primaryType;
}

// Nearby Search (New) rejects maxResultCount > 20.
export const PLACES_MAX_RESULTS = 20;

// ICP vertical → valid Google Place types (Table A). Inverse of
// TYPE_NORMALIZATION, minus types the API rejects (store, day_spa, boutique).
// The API 400s on unknown types, so only verified Table A types belong here.
export const VERTICAL_PLACE_TYPES = {
  restaurant: ["restaurant", "meal_takeaway", "meal_delivery", "pizza_restaurant", "mexican_restaurant"],
  bar: ["bar", "pub", "night_club"],
  cafe: ["cafe", "coffee_shop"],
  retail: ["clothing_store", "shoe_store", "gift_shop", "furniture_store"],
  salon: ["hair_salon", "beauty_salon", "nail_salon", "barber_shop"],
  spa: ["spa"],
  auto: ["car_repair", "car_wash", "auto_parts_store"],
  professional: ["dentist", "doctor", "lawyer", "accounting", "veterinary_care"],
};

// Expand ICP verticals into the deduped set of Google Place types for
// includedTypes. Unknown values pass through so a raw Google type still works.
export function verticalsToPlaceTypes(verticals) {
  const out = [];
  for (const v of verticals) {
    for (const t of VERTICAL_PLACE_TYPES[v] ?? [v]) {
      if (!out.includes(t)) out.push(t);
    }
  }
  return out;
}

export function parsePlacesResponse(raw) {
  const out = [];
  for (const p of raw.places ?? []) {
    const reviews = (p.reviews ?? [])
      .filter((r) => r.text)
      .map((r) => (r.text && r.text.text) || "");
    out.push(business({
      place_id: p.id,
      name: (p.displayName && p.displayName.text) || "",
      category: normalizeCategory(p.primaryType ?? ""),
      address: p.formattedAddress ?? "",
      phone: p.nationalPhoneNumber ?? null,
      website: p.websiteUri ?? null,
      rating: p.rating ?? null,
      review_count: p.userRatingCount ?? 0,
      price_level: PRICE_LEVELS[p.priceLevel] ?? null,
      business_status: p.businessStatus ?? "",
      review_texts: reviews.filter((t) => t),
    }));
  }
  return out;
}

export function dedupe(businesses, seenIds) {
  return businesses.filter((b) => !seenIds.has(b.place_id));
}

export function loadDemoBusinesses() {
  return parsePlacesResponse(demoRaw);
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// backoffMs is injectable so tests don't actually wait between retries.
export async function fetchNearby({ apiKey, location, radiusMeters, includedTypes,
  maxResults = 20, retries = 3, backoffMs = 1000 }) {
  const [lat, lng] = location.split(",").map(Number);
  const body = {
    includedTypes: verticalsToPlaceTypes(includedTypes),
    maxResultCount: Math.min(maxResults, PLACES_MAX_RESULTS),
    locationRestriction: { circle: { center: { latitude: lat, longitude: lng }, radius: Number(radiusMeters) } },
  };
  const headers = {
    "Content-Type": "application/json",
    "X-Goog-Api-Key": apiKey,
    "X-Goog-FieldMask": FIELD_MASK,
  };
  for (let attempt = 0; attempt < retries; attempt++) {
    const resp = await fetch(PLACES_URL, { method: "POST", headers, body: JSON.stringify(body) });
    if (resp.status === 429) {
      await sleep(2 ** attempt * backoffMs);
      continue;
    }
    if (!resp.ok) {
      const detail = await resp.text().catch(() => "");
      throw new Error(`Places API error ${resp.status}${detail ? `: ${detail}` : ""}`);
    }
    return parsePlacesResponse(await resp.json());
  }
  throw new Error("Places API rate limit: exhausted retries");
}
